### PR TITLE
fix(amplify_flutter): Add a new method isConfigured to Amplify

### LINKED ISSUE #300: "Auth plugin was not added" exception upon returning to screen where Amplify Auth plugin was initialised
**Describe the bug**
Thanks for the work you guys are doing here! I am getting an "**_Auth plugin was not added_**" exception upon returning to the screen where I initialised the Amplify Auth plugin. This is happening on `0.0.1-dev.6`

**To Reproduce**
1. Visit the initial screen where amplify auth plugin is configured.
2. Visit another screen
3. Return to the initial screen where amplify auth plugin is configured
4. See error.

**Expected behavior**
There should not be an error displayed.

**Screenshots**
Here's what the error looks like:

```
flutter: Auth plugin was not added
[VERBOSE-2:ui_dart_state.cc(177)] Unhandled Exception: Amplify plugin was not added
#0      Amplify.addPlugin (package:amplify_core/amplify_core.dart:72:9)
#1      _EntryScreenState._configureAmplify (package:amplify_app_demo/screens/entry.dart:29:23)
#2      _EntryScreenState.initState (package:amplify_app_demo/screens/entry.dart:21:5)
#3      StatefulElement._firstBuild (package:flutter/src/widgets/framework.dart:4821:57)
#4      ComponentElement.mount (package:flutter/src/widgets/framework.dart:4658:5)
#5      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3624:14)
#6      Element.updateChild (package:flutter/src/widgets/framework.dart:3389:18)
#7      SingleChildRenderObjectElement.mount (package:flutter/src/widgets/framework.dart:6173:14)
#8      Element.inflateWidget (package:flutter/src/widgets/framework.dart:3624:14)
#9      Element.updateChild (package:flutter/src/widgets/framework.dart:3389:18)
#10     ComponentElement.performRebuild (package:flutter/src/widgets/framework<…>
```

**Platform**
Amplify Flutter current supports iOS and Android.  This issue is reproducible in (check all that apply):
[x] Android
[x] iOS

**Smartphone (please complete the following information):**
 - Device: iPhone 12 Pro Max Simulator
 - OS: iOS 14.3

**Additional context**
I reckon this happens due to the static Auth property that gets initialised on the `Amplify` class:

```dart
// amplify_core.dart
class Amplify {
  static const AuthCategory Auth = const AuthCategory();
...
}
```
The `AuthCategory` instance also has a static list contains the plugins configured:

```dart
// amplify_auth_category.dart
class AuthCategory {
...
  static List<AuthPluginInterface> plugins = [];
...
}
```
Since they are treated as singletons, whenever the `addPlugin()` method is invoked on `Amplify()` the following conditions trigger a failure:

```dart
// amplify_core.dart
Future<void> addPlugin({List<AuthPluginInterface> authPlugins,...}) async {
    if (!_isConfigured) {
      try {
        if (authPlugins != null && authPlugins.length == 1) {
          Auth.addPlugin(authPlugins[0]); // <-- 
        }
  ...
}

// amplify_auth_category.dart
class AuthCategory {
...
  void addPlugin(AuthPluginInterface plugin) {
      if (plugins.length == 0) {
        plugins.add(plugin);
        
      } else {
        throw("Auth plugin was not added"); // <-- This happens since the plugin was added earlier 
      }
    }
...
}
```

I came across this related issue https://github.com/aws-amplify/amplify-flutter/issues/124 where it's suggested to check that the auth plugin is already configured. I implemented this extension method as a temporary workaround:

```dart
import 'package:amplify_core/amplify_core.dart';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_analytics_pinpoint/amplify_analytics_pinpoint.dart';
import 'package:amplify_core_plugin_interface/amplify_core_plugin_interface.dart'; // <--

extension on Amplify {
  bool hasConfiguredAuth() => AuthCategory.plugins.isNotEmpty;
}

// and using like so:
if (!amplifyInstance.hasConfiguredAuth()) {
  amplifyInstance.addPlugin(
    authPlugins: [auth],
    analyticsPlugins: [analytics],
  );
}
```

Keep up the great work! Looking to do a tutorial series about Amplify Flutter on my YouTube channel soon 😎 https://youtube.com/c/CreativeBracket

### PR DESCRIPTION
**Pull request by Amplifiyer**
Several other enhancements including error messages, documentation and unit tests.

*Issue #, if available:* Fixes #300, #302, #124

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_flutter/lib/amplify.dart [hunk 1]
```diff
@@ -68,29 +79,48 @@ class AmplifyClass extends PlatformInterface {
...
+  /// to check the configuration state by calling `Amplify.isConfigured()`
   Future<void> addPlugins(List<AmplifyPluginInterface> plugins) async {
     plugins.forEach((plugin) async {
       await addPlugin(plugin);
     });
     return;
   }
 
+  /// Returns whether Amplify has been configured or not.
+  bool isConfigured() {
```
Comment (haverchuck) [inline packages/amplify_flutter/lib/amplify.dart:98]: Perhaps make this a getter?

**Review by haverchuck — CHANGES_REQUESTED** [review]
I would suggest making `isConfigured` a getter - unless there is a good reason not to do so.

File: packages/amplify_flutter/lib/amplify.dart (hunk 1 above)
Comment (Amplifiyer) [reply packages/amplify_flutter/lib/amplify.dart]: Done

**Review by haverchuck — APPROVED** [review]
I would suggest making `isConfigured` a getter - unless there is a good reason not to do so.

**Pull request merged**